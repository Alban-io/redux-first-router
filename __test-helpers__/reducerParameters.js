import createSmartHistory from '../src/smart-history'

import { createInitialState } from '../src/createLocationReducer'

export default async (type, pathname) => {
  // eslint-disable-line import/prefer-default-export
  const history = createSmartHistory({ initialEntries: ['/first'] })
  history.listen(function() {})
  await history.push(pathname)

  const current = { pathname, type, payload: { param: 'bar' } }
  const prev = { pathname: '/first', type: 'FIRST', payload: {} }
  const routesMap = {
    FIRST: { path: '/first' },
    SECOND: { path: '/second/:param' }
  }

  return {
    type,
    pathname,
    current,
    prev,

    initialState: createInitialState(
      routesMap,
      history
    ),

    routesMap,
    history,

    action: {
      type,
      payload: { param: 'bar' },
      meta: {
        location: {
          current,
          prev,
          kind: 'load',
          history: {
            entries: history.entries.slice(0), // history.entries.map(entry => entry.pathname)
            index: history.index,
            length: history.length
          }
        }
      }
    },

    expectState(state) {
      expect(state.pathname).toEqual(pathname)
      expect(state.type).toEqual(type)
      expect(state.payload).toEqual({ param: 'bar' })
      expect(state.prev).toEqual(prev)
      expect(state.kind).toEqual('load')

      expect(state).toMatchSnapshot()
    }
  }
}
