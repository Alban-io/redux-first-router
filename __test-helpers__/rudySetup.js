import { applyMiddleware, createStore, compose, combineReducers } from 'redux'
import reduxThunk from 'redux-thunk'
import createRouter from '../src/createRouter'
import createScene from '../src/createScene'
import { NOT_FOUND } from '../src/index'

import fakeAsyncWork from '../__test-helpers__/fakeAsyncWork'

export default (path = '/first', options = {}, custom = {}) => {
  const routesMap = {
    ...defaultRoutes,
    ...custom.routesMap
  }

  options.initialEntries = [path]
  options.extra = { arg: 'extra-arg' }

  const { types, actions, routes, exportString } = createScene(routesMap, custom)
  const { middleware, reducer, firstRoute, history } = createRouter(
    routes,
    options
  )

  const title = (state = {}, action = {}) => action.type
  const rootReducer = combineReducers({ title, location: reducer })
  const enhancer = applyMiddleware(middleware, reduxThunk)
  const store = createStore(rootReducer, enhancer)

  return {
    store,
    firstRoute,
    history,
    types,
    actions,
    exportString,
    routes
  }
}

export const defaultRoutes = {
  FIRST: '/first',
  SECOND: '/second',
  THIRD: {
    path: '/third',
    thunk: async () => {
      await fakeAsyncWork()
      return 'thunk'
    }
  },
  FOURTH: {
    path: '/fourth',
    thunk: async () => {
      await fakeAsyncWork()
      return 'thunk'
    },
    onComplete: () => {
      return 'onComplete'
    },
    action: (arg) => (req, type) => {
      return { payload: { foo: arg }, type }
    }
  },
  PLAIN: {
    action: (arg) => {
      return { foo: arg }
    }
  },
  [NOT_FOUND]: '/not-found-foo'
}

export const log = store => {
  const state = store.getState().location
  delete state.routesMap
  delete state.hasSSR
  console.log(state)
}
