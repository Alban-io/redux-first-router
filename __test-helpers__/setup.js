import createHistory from 'history/createMemoryHistory'
import connectRoutes from '../src/connectRoutes'


export default (path = '/', options = { title: 'title', location: 'location' }) => {
  const routesMap = {
    FIRST: '/first',
    SECOND: '/second/:param',
  }

  const history = createHistory({
    initialEntries: [path],
    initialIndex: 0,
    keyLength: 6,
  })

  return connectRoutes(history, routesMap, options)
}
