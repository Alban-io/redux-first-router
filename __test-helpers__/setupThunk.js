import { createStore, applyMiddleware, compose } from 'redux'
import createHistory from 'history/createMemoryHistory'
import connectTypes from '../src/connectTypes'


export default (path = '/', thunk) => {
  const routesMap = {
    FIRST: '/first',
    SECOND: { path: '/second/:param', thunk },
    THIRD: { path: '/third/:param' },
  }

  const history = createHistory({
    initialEntries: [path],
    initialIndex: 0,
    keyLength: 6,
  })

  const { middleware, enhancer, thunk: ssrThunk, reducer: locationReducer } = connectTypes(history, routesMap)

  const middlewares = applyMiddleware(middleware)

  const reducer = (state = {}, action = {}) => ({
    location: locationReducer(state.location, action),
  })

  const store = createStore(reducer, undefined, compose(enhancer, middlewares))

  return { store, thunk: ssrThunk }
}
