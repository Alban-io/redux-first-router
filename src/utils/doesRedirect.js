import isRedirect from './isRedirect'

const noOp = function() {}

export default (action, redirectFunc = noOp) => {
  if (isRedirect(action)) {
    const url = action.location.url
    const status = action.location.status || 302
    redirectFunc(status, url, action)
    return true
  }
}
