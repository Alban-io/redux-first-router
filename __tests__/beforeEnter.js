import { setupAll } from '../__test-helpers__/setup'
import fakeAsyncWork from '../__test-helpers__/fakeAsyncWork'
import tempMock from '../__test-helpers__/tempMock'
import doesRedirect from '../src/utils/doesRedirect'

it('calls beforeEnter handler on route change -- route', async () => {
  const beforeEnter = jest.fn()

  const routesMap = {
    FIRST: '/first',
    SECOND: {
      path: '/second',
      beforeEnter
    },
    THIRD: '/third'
  }

  const { store } = await setupAll('/first', undefined, { routesMap })

  const action = { type: 'SECOND' }
  await store.dispatch(action)

  expect(beforeEnter).toHaveBeenCalled()
  expect(beforeEnter.mock.calls[0][0].action).toMatchObject(action)
  expect(beforeEnter.mock.calls[0][0].arg).toEqual('extra-arg')
})

it('calls beforeEnter handler on route change -- global', async () => {
  const beforeEnter = jest.fn()
  const { store } = await setupAll('/first', { beforeEnter })

  const action = { type: 'SECOND', payload: { param: 'bar' } }
  await store.dispatch(action)

  expect(beforeEnter).toHaveBeenCalled()
  expect(beforeEnter.mock.calls[1][0].action).toMatchObject(action)
  expect(beforeEnter.mock.calls[1][0].arg).toEqual('extra-arg')
})

it('if beforeEnter dispatches redirect, route changes with kind === "redirect"', async () => {
  const beforeEnter = jest.fn(({ dispatch, getState, action }) => {
    if (action.type !== 'SECOND') return
    return { type: 'THIRD' } // returned plain objects automatically dispatched
  })

  const { store, history } = await setupAll('/first', { beforeEnter })
  const action = await store.dispatch({ type: 'SECOND', payload: { param: 'bar' } })
  const { location } = store.getState()

  expect(action.type).toEqual('THIRD')

  expect(location.kind).toEqual('redirect')
  expect(location.type).toEqual('THIRD')
  expect(history.entries.length).toEqual(2)
  expect(location).toMatchSnapshot()
  expect(beforeEnter.mock.calls.length).toEqual(3)
})

it('beforeEnter redirect on server results does not update state, and instead returns action (i.e. short-circuits)', async () => {
  tempMock('../src/utils/isServer', () => () => true)
  const { setupAll } = require('../__test-helpers__/setup')

  const beforeEnter = jest.fn(({ dispatch, getState, action }) => {
    if (action.type !== 'SECOND') return
    dispatch({ type: 'THIRD' })
  })

  const { store, history } = await setupAll('/first', {
    beforeEnter
  })

  const action = await store.dispatch({
    type: 'SECOND',
    payload: { param: 'bar' }
  })

  const { location } = store.getState()
  expect(history.entries.length).toEqual(1)
  expect(action.kind).toEqual('redirect')
  expect(action.type).toEqual('THIRD')
  expect(action.meta.location.status).toEqual(302)
  expect(action.meta.location.url).toEqual('/third')

  const redirect = jest.fn()
  expect(doesRedirect(action, redirect)).toEqual(true)
  expect(redirect).toBeCalledWith(302, '/third')

  expect(location.type).toEqual('FIRST')
  expect(location).toMatchSnapshot()
  expect(action).toMatchSnapshot()
})

it('beforeEnter delays route changes until beforeEnter promise resolves', async () => {
  const beforeEnter = jest.fn(async ({ dispatch, getState, action }) => {
    if (action.type !== 'SECOND') return

    await fakeAsyncWork()
    dispatch({ type: 'NON_ROUTE_AUTH_ACTION' })
  })

  const routesMap = {
    FIRST: '/first',
    SECOND: {
      path: '/second',
      thunk: async ({ dispatch, getState }) => {
        await fakeAsyncWork()
        return 'thunkReturn'
      }
    }
  }

  const { store } = await setupAll('/first', { beforeEnter }, { routesMap })

  const prom = store.dispatch({ type: 'SECOND' })
  expect(store.getState().location.type).toEqual('FIRST')

  const res = await prom
  expect(res.payload).toEqual('thunkReturn')
  expect(res.type).toEqual('SECOND_COMPLETE')
  expect(store.getState().location.type).toEqual('SECOND')
  expect(store.getState().location.kind).toEqual('push')

  // VERY IMPORTANT:
  // there will be 4 dispatches, but only 2 calls to `beforeEnter`:
  // - 1 initialDispatch in setupAll
  // - 2 externally by the user (SECOND, SOME_NON_ROUTE_TYPE_SUCH_AS_AUTH_STUFF)
  // - and 1 additional by the middleware to re-dispatch SECOND
  expect(beforeEnter.mock.calls.length).toEqual(2)
})

it('beforeEnter that returns a promise which redirects', async () => {
  const beforeEnter = jest.fn(async ({ dispatch, getState, action }) => {
    if (action.type !== 'SECOND') return

    await fakeAsyncWork()
    return { type: 'THIRD' } // returned plain objects automatically dispatched
  })

  const routesMap = {
    FIRST: '/first',
    SECOND: '/second',
    THIRD: {
      path: '/third',
      thunk: async (dispatch, getState) => {
        await fakeAsyncWork()
        return 'beforeEnterPlusThunkReturn'
      }
    }
  }

  const { store } = await setupAll('/first', { beforeEnter }, { routesMap })

  expect(store.getState().location.type).toEqual('FIRST')

  const res = await store.dispatch({ type: 'SECOND' })
  expect(res.payload).toEqual('beforeEnterPlusThunkReturn')
  expect(res.type).toEqual('THIRD_COMPLETE')
  expect(store.getState().location.type).toEqual('THIRD')
  expect(store.getState().location.kind).toEqual('redirect')

  // VERY IMPORTANT:
  // we expect `beforeEnter` to be called 3 times, because even the final/3rd
  // route redirected to must be filtered again. In the final route, there are no more
  // redirects so it passes the filtering.
  expect(beforeEnter.mock.calls.length).toEqual(3)
})
