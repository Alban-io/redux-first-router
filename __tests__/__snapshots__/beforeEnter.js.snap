// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`beforeEnter redirect on server results does not update state, and instead returns action (i.e. short-circuits) 1`] = `
Object {
  "basename": "",
  "entries": Array [
    Object {
      "hash": "",
      "key": "345678",
      "pathname": "/first",
      "search": "",
      "state": Object {},
      "url": "/first",
    },
  ],
  "hasSSR": true,
  "index": 0,
  "kind": "load",
  "length": 1,
  "pathname": "/first",
  "payload": Object {},
  "prev": Object {
    "index": -1,
    "kind": "",
    "length": 0,
    "pathname": "",
    "payload": Object {},
    "type": "",
  },
  "type": "FIRST",
  "url": "/first",
}
`;

exports[`beforeEnter redirect on server results does not update state, and instead returns action (i.e. short-circuits) 2`] = `
Object {
  "kind": "redirect",
  "meta": Object {
    "location": Object {
      "committed": false,
      "current": null,
      "kind": "redirect",
      "status": 302,
      "url": "/third",
    },
  },
  "type": "THIRD",
}
`;

exports[`if beforeEnter dispatches redirect, route changes with kind === "redirect" 1`] = `
Object {
  "basename": "",
  "entries": Array [
    Object {
      "hash": "",
      "key": "345678",
      "pathname": "/first",
      "search": "",
      "state": Object {},
      "url": "/first",
    },
    Object {
      "hash": "",
      "key": "345678",
      "pathname": "/third",
      "search": "",
      "state": Object {},
      "url": "/third",
    },
  ],
  "hasSSR": undefined,
  "index": 1,
  "kind": "redirect",
  "length": 2,
  "pathname": "/third",
  "payload": Object {},
  "prev": Object {
    "index": 1,
    "kind": "push",
    "length": 2,
    "pathname": "/second/bar",
    "payload": Object {
      "param": "bar",
    },
    "search": "",
    "type": "SECOND",
    "url": "/second/bar",
  },
  "type": "THIRD",
  "url": "/third",
}
`;

exports[`route.redirect - automatic 301 redirect 1`] = `
Object {
  "basename": "",
  "entries": Array [
    Object {
      "hash": "",
      "key": "345678",
      "pathname": "/first",
      "search": "",
      "state": Object {},
      "url": "/first",
    },
    Object {
      "hash": "",
      "key": "345678",
      "pathname": "/third",
      "search": "",
      "state": Object {},
      "url": "/third",
    },
  ],
  "hasSSR": undefined,
  "index": 1,
  "kind": "redirect",
  "length": 2,
  "pathname": "/third",
  "payload": Object {
    "param": "bar",
  },
  "prev": Object {
    "index": 1,
    "kind": "push",
    "length": 2,
    "pathname": "/second/bar",
    "payload": Object {
      "param": "bar",
    },
    "search": "",
    "type": "SECOND",
    "url": "/second/bar",
  },
  "type": "THIRD",
  "url": "/third",
}
`;
